### ISSUE TITLE
Inconsistencies in IdnMapping on Unix and Windows

### ISSUE THREAD

**Issue by hughbe** [dotnet/corefx#8242]
Originally, some of these were found in #7921 and I found more inconsistencies in #8224 so its worth creating a new issue and closing the old one that diagnosed the problems as with Uri.GetComponents.
## GetAscii - uppercase ASCII chars (e.g. `A, B, C`)
- Windows behaviour: returns the uppercase chars
- Unix behaviour: returns the lowercase chars
## GetAscii - single dash (e.g. `-`)
- Windows behaviour: returns the original string
- Unix behaviour: throws an `ArgumentException`
## GetAscii - single full stop (e.g. `.`)
- Windows behaviour: throws an `ArgumentException`
- Mac OSX behaviour: throws an `ArgumentException`
- Unix behaviour: returns the original string
## GetUnicode - invalid punycode (e.g. `xn--\u1234pck`)
- Windows behaviour: returns the original string
- Unix behaviour: returns `ఽూ`

/cc @ellismg @tarekgh

**Comment by tarekgh** [comment]
@ellismg do you have more info about this issue?

**Comment by ellismg** [comment]
I will need to dig in and consult the IDNA specs.  Last time I talk to Shawn about some of these it sounded like Windows may not follow the spec exactly where as ICU did and that's why we saw differences in some cases.

**Comment by tarekgh** [comment]
@ellismg the question is, should we always give the OS behavior (as we do for the sorting)? or should we try to always follow the specs and have consistent behavior? 

we already have differences between Windows and Linux in some areas. Idn could be another one here.

**Comment by hughbe** [comment]
Just a quick update: I filed a bug report over at the ICU project for `GetUnicode("xn--\u1234pck")` and its been accepted. This is an example of where ICU doesn't follow the IDNA spec.

@ellismg if you find any other places ICU doesn't follow the spec in your digging, I can file a bug and get it fixed for future ICU versions.

Here's the bug http://bugs.icu-project.org/trac/ticket/12523

**Comment by ellismg** [comment]
> Just a quick update: I filed a bug report over at the ICU project for GetUnicode("xn--\u1234pck") and its been accepted. This is an example of where ICU doesn't follow the IDNA spec.

> Here's the bug http://bugs.icu-project.org/trac/ticket/12523

Nice!

**Comment by tarekgh** [comment]
I have spent some time looking at the Idn specs and talking to some people offline and here is the conclusion of the issue discussed here:

> GetAscii - uppercase ASCII chars (e.g.  A, B, C )

Windows doesn't touch the characters if it is ASCII but looks on Linux it lowercase the ASCII characters. although Windows behavior seems more correct according to spec when doing ToAscii without passing any flags, the ASCII characters are not touched. but Linux results can be correct too as the specs states the following 

```
Labels are defined to be equivalent if and only if their ASCII forms produced by ToASCII match using a case-insensitive ASCII comparison.
```

> GetAscii - single dash (e.g.  - )

According to the specs Windows behavior is more correct here as it just output "-" especially there is no passed flags like UseSTD3ASCIIRules. on Linux it throw but again this can make sense too as "-" by itself is not really valid label in the real world scenario. we can log issue to Linux but I am really seeing this is can be seen as by design too there.

> GetAscii - single full stop (e.g.  . )

This is similar issue as the previous point but it is opposite behaviors on Windows and Linux. on Windowsand OSX it throws but in Linux it produce ".". Linux behavior more correct according to the specs. I talked wo Windows team and they said this is not interesting case because "." by itself is not correct label to throwing behavior is acceptable here.

> GetUnicode - invalid punycode (e.g.  xn--\u1234pck )

i tis stated in this issue that on Windows we return the same string. but after testing it on Windows I am seeing we throw (as expected). @hughbe already logged issue for ICU for the wrong results returned from such call. 

To conclude the issue:
it is expected we can get some differences when running on Windows than we running on OSX or Linux. the bottom line is these differences are acceptable and our tests should handle such differences. By that I am closing this issue as there is no real action item here.
